File: modules/heaps/heaps/linkedheap/LinkedHeap.java
package heaps.linkedheap;

import heaps.linkedheap.exceptions.*;

/**
 * LinkedHeap implements a heap.
 *
 * @author Java Foundations
 * @version 4.0
 */
public class LinkedHeap<T> extends LinkedBinaryTree<T> implements HeapADT<T> 
{
	public HeapNode<T> lastNode;  

	public LinkedHeap() 
	{
		super();
	}

	/**
	 * Adds the specified element to this heap in the appropriate
	 * position according to its key value. 
	 *
	 * @param obj the element to be added to the heap
	 */
	public void addElement(T obj) 
	{
		HeapNode<T> node = new HeapNode<T>(obj);

		if (root == null)
			root=node;
		else
		{
			HeapNode<T> nextParent = getNextParentAdd(); // keep heap structure
			// Determine which child
			if (nextParent.getLeft() == null)  // no children
				nextParent.setLeft(node);
			else                               // one child - the left node
				nextParent.setRight(node);

			node.setParent(nextParent);
		}
		lastNode = node;
		modCount++;

		if (size() > 1)
			heapifyAdd();  // keep the heap ordering
	}

	/**
	 * Returns the node that will be the parent of the new node
	 *
	 * @return the node that will be the parent of the new node
	 */
	private HeapNode<T> getNextParentAdd()
	{
		HeapNode<T> result = lastNode;

		// Get to the left sub-tree or the root
		while ((result != root) && (result.getParent().getLeft() != result))
			result = result.getParent();
		
		if (result != root)
			// Go to the parent's right subtree
			if (result.getParent().getRight() == null)
				// Parent has no right child, so parent is the new parent
				result = result.getParent();
			else
			{
				// Get parent's right child
				result = (HeapNode<T>)result.getParent().getRight();
				// Go all the way to the left
				while (result.getLeft() != null)
					result = (HeapNode<T>)result.getLeft();
			}
		else
			// Tree is full so go all the way to the left (start a new row of leaves)
			while (result.getLeft() != null)
				result = (HeapNode<T>)result.getLeft();

		return result;
	}

	/**
	 * Reorders this heap after adding a node.
	 */
	private void heapifyAdd()
	{
		T temp;
		HeapNode<T> next = lastNode;

		temp = next.getElement();

		while ((next != root) && 
				(((Comparable<T>)temp).compareTo(next.getParent().getElement()) < 0))
		{
			next.setElement(next.getParent().getElement());
			next = next.parent;
		}
		next.setElement(temp);
	}

	/**
	 * Remove the element with the lowest value in this heap and
	 * returns a reference to it. Throws an EmptyCollectionException 
	 * if the heap is empty.
	 *
	 * @return the element with the lowest value in this heap
	 * @throws EmptyCollectionException if the heap is empty
	 */
	public T removeMin() throws EmptyCollectionException 
	{
		if (isEmpty())
			throw new EmptyCollectionException("LinkedHeap");

		T minElement =  root.getElement();

		// If the size is 1 the heap will be empty
		if (size() == 1)
		{
			root = null;
			lastNode = null;
		}
		else
		{
			// Get the new last node
			HeapNode<T> newLast = getNewLastNode();
			
			// Check which child the old last node is & set to null
			if (lastNode.getParent().getLeft() == lastNode)
				lastNode.getParent().setLeft(null);
			else
				lastNode.getParent().setRight(null);

			((HeapNode<T>)root).setElement(lastNode.getElement());
			lastNode = newLast;
			heapifyRemove();
		}
		modCount++;
		return minElement;
	}

	/**
	 * Reorders this heap after removing the root element.
	 */
	private void heapifyRemove()
	{
		T temp;
		HeapNode<T> node = (HeapNode<T>)root;
		HeapNode<T> left = (HeapNode<T>)node.getLeft();
		HeapNode<T> right = (HeapNode<T>)node.getRight();
		HeapNode<T> next;

		
		if ((left == null) && (right == null))
			// no children
			next = null;
		else if (right == null)
			// one child
			next = left;
		else if (((Comparable<T>)left.getElement()).compareTo(right.getElement()) < 0)
			// two children, left is smaller
			next = left;
		else
			// two children, right is smaller
			next = right;

		temp = node.getElement();
		while ((next != null) && 
				(((Comparable<T>)next.getElement()).compareTo(temp) < 0))
		{
			node.setElement(next.getElement());
			node = next;
			left = (HeapNode<T>)node.getLeft();
			right = (HeapNode<T>)node.getRight();

			if ((left == null) && (right == null))
				next = null;
			else if (right == null)
				next = left;
			else if (((Comparable<T>)left.getElement()).compareTo(right.getElement()) < 0)
				next = left;
			else
				next = right;
		}
		node.setElement(temp);
	}

	/**
	 * Returns the node that will be the new last node after a remove.
	 *
	 * @return the node that will be the new last node after a remove
	 */
	private HeapNode<T> getNewLastNode()
	{
		HeapNode<T> newLastNode = lastNode;

		
		while ((newLastNode != root) && (newLastNode.getParent().getLeft() == newLastNode))
			newLastNode = newLastNode.getParent();

		if (newLastNode != root)
			newLastNode = (HeapNode<T>)newLastNode.getParent().getLeft();

		while (newLastNode.getRight() != null)
			newLastNode = (HeapNode<T>)newLastNode.getRight();

		return newLastNode;
	}

	/**
	 * Returns the element with the lowest value in this heap.
	 * Throws an EmptyCollectionException if the heap is empty.
	 *
	 * @return the element with the lowest value in this heap
	 * @throws EmptyCollectionException if the heap is empty
	 */
	public T findMin() throws EmptyCollectionException
	{
		if (isEmpty())
			throw new EmptyCollectionException("LinkedHeap");

		return root.getElement();
	}
	
}

File: modules/heaps/python/linked_heap.py
"""
LinkedHeap implements a min heap on top of a LinkedBinaryTree.

1-1 conversion of heaps/linkedheap/LinkedHeap.java
(Java's Comparable.compareTo(x) < 0 becomes Python's "a < b").
"""

from linked_binary_tree import LinkedBinaryTree
from heap_node import HeapNode
from exceptions import EmptyCollectionException


class LinkedHeap(LinkedBinaryTree):

    def __init__(self):
        super().__init__()
        self.last_node = None

    def add_element(self, obj):
        # Adds the specified element to this heap in the appropriate position.
        node = HeapNode(obj)

        if self.root is None:
            self.root = node
        else:
            next_parent = self.get_next_parent_add()  # keep heap structure
            # Determine which child
            if next_parent.get_left() is None:    # no children
                next_parent.set_left(node)
            else:                                 # one child - the left node
                next_parent.set_right(node)

            node.set_parent(next_parent)
        self.last_node = node
        self.mod_count += 1

        if self.size() > 1:
            self.heapify_add()  # keep the heap ordering

    def get_next_parent_add(self):
        # Returns the node that will be the parent of the new node.
        result = self.last_node

        # Get to the left sub-tree or the root
        while (result is not self.root) and (result.get_parent().get_left() is not result):
            result = result.get_parent()

        if result is not self.root:
            # Go to the parent's right subtree
            if result.get_parent().get_right() is None:
                # Parent has no right child, so parent is the new parent
                result = result.get_parent()
            else:
                # Get parent's right child
                result = result.get_parent().get_right()
                # Go all the way to the left
                while result.get_left() is not None:
                    result = result.get_left()
        else:
            # Tree is full so go all the way to the left (start a new row of leaves)
            while result.get_left() is not None:
                result = result.get_left()

        return result

    def heapify_add(self):
        # Reorders this heap after adding a node.
        next = self.last_node

        temp = next.get_element()

        while (next is not self.root) and (temp < next.get_parent().get_element()):
            next.set_element(next.get_parent().get_element())
            next = next.parent
        next.set_element(temp)

    def remove_min(self):
        # Remove and return the element with the lowest value in this heap.
        if self.is_empty():
            raise EmptyCollectionException("LinkedHeap")

        min_element = self.root.get_element()

        # If the size is 1 the heap will be empty
        if self.size() == 1:
            self.root = None
            self.last_node = None
        else:
            # Get the new last node
            new_last = self.get_new_last_node()

            # Check which child the old last node is & set to null
            if self.last_node.get_parent().get_left() is self.last_node:
                self.last_node.get_parent().set_left(None)
            else:
                self.last_node.get_parent().set_right(None)

            self.root.set_element(self.last_node.get_element())
            self.last_node = new_last
            self.heapify_remove()
        self.mod_count += 1
        return min_element

    def heapify_remove(self):
        # Reorders this heap after removing the root element.
        node = self.root
        left = node.get_left()
        right = node.get_right()

        if (left is None) and (right is None):
            # no children
            next = None
        elif right is None:
            # one child
            next = left
        elif left.get_element() < right.get_element():
            # two children, left is smaller
            next = left
        else:
            # two children, right is smaller
            next = right

        temp = node.get_element()
        while (next is not None) and (next.get_element() < temp):
            node.set_element(next.get_element())
            node = next
            left = node.get_left()
            right = node.get_right()

            if (left is None) and (right is None):
                next = None
            elif right is None:
                next = left
            elif left.get_element() < right.get_element():
                next = left
            else:
                next = right
        node.set_element(temp)

    def get_new_last_node(self):
        # Returns the node that will be the new last node after a remove.
        new_last_node = self.last_node

        while (new_last_node is not self.root) and (new_last_node.get_parent().get_left() is new_last_node):
            new_last_node = new_last_node.get_parent()

        if new_last_node is not self.root:
            new_last_node = new_last_node.get_parent().get_left()

        while new_last_node.get_right() is not None:
            new_last_node = new_last_node.get_right()

        return new_last_node

    def find_min(self):
        # Returns the element with the lowest value in this heap.
        if self.is_empty():
            raise EmptyCollectionException("LinkedHeap")

        return self.root.get_element()
